### ISSUE TITLE
Issue with 17c80c02 and point color with colors

### ISSUE THREAD

**Issue by the-glu** [PointCloudLibrary/pcl#1324]
Hello,

I have an issue with my RGB points cloud, colors are wrong on the master branch.

With a `git bisec` the issue started with 17c80c02 , but I wasen't able to fix the problem myself.

I load my point cloud with `pcl::io::loadPCDFile<pcl::PointXYZRGB>` and display it in a `PCLVisualizer` with `viewer->addPointCloud([A pcl::PointCloud<pcl::PointXYZRGB>::Prt])`.

I noticed in the commit that the last changes are probably wrong ( https://github.com/PointCloudLibrary/pcl/commit/17c80c025dc2a9e174df5ac072ccb37081359e47#diff-a2909a988624f3dcc1638a094d9add63L179 ) (using the `j` variable instead of `idx`), but that dosen't fix my problem.

`rgba_index` is equal to 3.

PCD file look like this:

```
# .PCD v0.7 - Point Cloud Data file format
VERSION 0.7
FIELDS x y z rgb
SIZE 4 4 4 4
TYPE F F F F
COUNT 1 1 1 1
WIDTH 55190
HEIGHT 1
VIEWPOINT 0 0 0 1 0 0 0
POINTS 55190
DATA ascii
-0.25698575 -0.37610915 1.6564335 2.1482638e-038
-0.25410366 -0.37628013 1.6570516 2.1482638e-038
(..)
```

**Comment by taketwo** [comment]
Hi, thanks for reporting this and sorry for the inconvenience.

Indeed, `j` is used by mistake instead of `idx`. This should not affect you though because your cloud has `xyz` and therefore the other `if` branch is executed.

Could you write more specifically what is wrong with colors? (But whatever this problem is, I suspect it has something to do with the fact that in your PCD file the `rgb` field is represented with a single float, whereas in code with copy RGB components byte-by-byte.)

**Comment by the-glu** [comment]
Point cloud look like this: ![https://trello-attachments.s3.amazonaws.com/55eda612f628532e8f108b44/577x350/a8e639c118d4cab340ae3dc169da1855/upload_9_8_2015_at_10_00_36_AM.png](https://trello-attachments.s3.amazonaws.com/55eda612f628532e8f108b44/577x350/a8e639c118d4cab340ae3dc169da1855/upload_9_8_2015_at_10_00_36_AM.png)

Indeed it's because I have float and not chars in my point for colors. Switching the line to `memcpy (&rgb, (reinterpret_cast<const float *> (&cloud_->points[cp])) + rgba_index * sizeof(float), sizeof (pcl::RGB));` fix the problem, however I don't know how to fix this properly ^^

**Comment by taketwo** [comment]
Looking at your fix it became obvious for me that the problem was not related to `float`/`char` conversions. But accidentally you hit the right spot by multiplying index with `sizeof(float)`.

The thing is, `rgba_index` is not a byte offset inside the point structure, but rather a field offset. As such, it should not be added to a `char` pointer. Instead, we should add `fields[rgba_index].offset`, which is a proper byte offset. (In our case, it is equal 16 bytes.)

Your fix does not actually do the right thing (although luckily it gives visually "correct" results). You add offset (`rgba_index * sizeof(float) = 3 * 4 = 12`) to a `float` pointer, effectively making (`12 * sizeof(float) = 48`) bytes offset. Since the size of `PointXYZRGB` is 32 bytes, this ends up offset 16 bytes with respect to the next point, which is exactly it's `rgb` field.

PR fixing both found bugs is welcome :)

**Comment by the-glu** [comment]
Okay, PR openned, thank for your help !

**Comment by taketwo** [comment]
Fixed via #1325
